service/storage_service.cc

```diff
@@ -3049,7 +3049,9 @@ future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmpt
     std::vector<std::unordered_map<sstring, locator::vnode_effective_replication_map_ptr>> pending_effective_replication_maps;
     pending_effective_replication_maps.resize(smp::count);
     std::vector<std::unordered_map<table_id, locator::effective_replication_map_ptr>> pending_table_erms;
```

Comment:
> Currently, the loop updating all tables (including views) with the
> new effective_replication_map may yield, and therefore expose
> a state where the base and view tables effective_replication_map
> and topology are out of sync (as seen in https://github.com/scylladb/scylladb/issues/17786)
> 
> To prevent that, loop over all base tables and for each table
> update the base table and all views atomically, without yielding,
> and so allow yielding only between base tables.

This solution works, but it's fragile: we recognize one interdependency in the schema, and plug that hole.

Cannot we capture the entire schema atomically, then work on it in our leisure? (typically a database schema refers to all tables, not just one).

Reply:
> > Currently, the loop updating all tables (including views) with the
> > new effective_replication_map may yield, and therefore expose
> > a state where the base and view tables effective_replication_map
> > and topology are out of sync (as seen in #17786)
> > To prevent that, loop over all base tables and for each table
> > update the base table and all views atomically, without yielding,
> > and so allow yielding only between base tables.
> 
> This solution works, but it's fragile: we recognize one interdependency in the schema, and plug that hole.
> 
> Cannot we capture the entire schema atomically, then work on it in our leisure? (typically a database schema refers to all tables, not just one).

It's possible using a prepare phase for calculating the modified state and then applying it atomically to all tables, but it is more complicated and higher risk, and so less backportable.
So I suggest starting with this simple approach and following up with a comprehensive fix going forward.